File: lib/features/provider_analytics/domain/provider_analytics.dart
class EngagementEvent {
  const EngagementEvent({
    required this.providerId,
    required this.opportunityId,
    required this.kind,
    required this.country,
    required this.studyLevel,
    required this.field,
    required this.occurredAt,
    this.userId,
    this.identifiableSharingConsent = false,
  });
  final String providerId;
  final String opportunityId;
  final String kind;
  final String country;
  final String studyLevel;
  final String field;
  final DateTime occurredAt;
  final String? userId;
  final bool identifiableSharingConsent;
}

class ProviderAnalyticsSnapshot {
  const ProviderAnalyticsSnapshot({
    required this.views,
    required this.saves,
    required this.applicationClicks,
    required this.countries,
    required this.studyLevels,
    required this.fields,
    required this.suppressed,
  });
  final int views;
  final int saves;
  final int applicationClicks;
  final Map<String, int> countries;
  final Map<String, int> studyLevels;
  final Map<String, int> fields;
  final bool suppressed;
}

abstract interface class ProviderAnalyticsRepository {
  Future<void> record(EngagementEvent event);
  Future<ProviderAnalyticsSnapshot> snapshot(String providerId);
  Future<String> exportCsv(String providerId);
}

File: scholarsphere_backend/app/schemas/provider_analytics.py
from datetime import datetime

from pydantic import BaseModel, Field


class RecordEngagementEventRequest(BaseModel):
    provider_id: str
    opportunity_id: str = Field(min_length=1, max_length=255)
    kind: str = Field(min_length=1, max_length=64)
    country: str = Field(min_length=1, max_length=255)
    study_level: str = Field(min_length=1, max_length=255)
    field: str = Field(min_length=1, max_length=255)
    occurred_at: datetime
    identifiable_sharing_consent: bool = False


class ProviderAnalyticsSnapshotRead(BaseModel):
    views: int
    saves: int
    application_clicks: int
    countries: dict[str, int]
    study_levels: dict[str, int]
    fields: dict[str, int]
    suppressed: bool


class ProviderAnalyticsExportRead(BaseModel):
    csv: str
